### ISSUE TITLE
Question/Discussion: Thoughts about Qt 6?

### ISSUE THREAD

**Issue by mojotx** [mrkite/TerraFirma#117]
TerraFirma has traditionally been built using a static build of [Qt 5](https://github.com/qt/qt5), but Qt is now up to version 6.0.3, and I wanted to know if there was any interest in moving to the new version, or if we want to continue to use a frozen version of Qt 5.

The installation instructions mention manually checking out the 5.12 tag, but I have always ignored that and used whatever recent version was available via Homebrew for macOS. However, now that I have Qt 6.0.3 installed, I need to either work on updating TerraFirma to support the newer versions of Qt, or else jump through the hoops mentioned to build a static version of Qt 5.12 for use by TerraFirma only. 

(I could download the release binary, but what's the fun in that? 😜)

Also, if we *DO* want to support Qt 6, do we want to retain support for older Qt5? 

For instance, in the file `l10n.cpp`, there's a problem with this function:

```c++
QList<QString> L10n::getLanguages() {
  return languages.toList();
}
```

With [Qt6, they've removed the `.toList()` method](https://doc.qt.io/qt-5/qset.html#toList), and you have to do something like this: 

```c++
QList<QString> L10n::getLanguages() {
  QList<QString> newList(languages.begin(), languages.end());
  return newList;
}
```

Would we want to add some precompiler checks to support both methods? Or do we want to do regression testing to figure out what new Qt6 things are backwards-compatible with Qt5?

**Comment by Faiakes** [comment]
I say we move it on to Qt 6.

**Comment by mrkite** [comment]
I am all for moving to Qt6.  Also, you'll notice that for the most recent windows binary, I switched to using the Qt Installer Framework and the stock Qt dlls.  Making a static build is such a pain in the ass that I just can't be bothered anymore.  So for the mac and windows binaries, let's use the latest Qt and use QTIF to make the binaries.

We do need to make sure that some version of Ubuntu is using Qt6 though, otherwise making linux binaries will become difficult.

**Comment by mojotx** [comment]
Awesome, moving to Qt6 is what I'd prefer as well. 

I will see how far I can get on it. Right now I'm trying to figure out why my compiler isn't finding the "QtOpenGLWidgets" header file, when I generate the Makefile with qmake.

**Comment by mrkite** [comment]
Another feature update alongside Qt6 is I would like to use the maintenance tool created by QtIF to automatically check for updates.

**Comment by mojotx** [comment]
I just submitted PR #118, if you would like to take a look. I tried not to affect any functionality, and just port the code to support new function names. 

I am not a Qt expert, so don't be shy about pointing out improvements. 

It does seem to work, and once I merged in the 3.1.4 release that @mrkite released over the weekend, I was able to load a Terraria 1.4.2.1 save game successfully.
